libraries/state_history/include/eosio/state_history/log.hpp

```diff
@@ -68,818 +66,405 @@ static const uint16_t ship_feature_pruned_log = 1;
+      log.punch_hole(fc::raw::pack_size(log_header()), prune_to_pos);
 
       _begin_block = prune_to_num;
-      log.flush();
+      ilog("${name} pruned to blocks ${b}-${e}", ("name", log.display_path())("b", _begin_block)("e", _end_block - 1));
+   }
 
-      if(auto l = fc::logger::get(); l.is_enabled(loglevel))
-         l.log(fc::log_message(fc::log_context(loglevel, __FILE__, __LINE__, __func__),
-                               "${name}.log pruned to blocks ${b}-${e}", fc::mutable_variant_object()("name", name)("b", _begin_block)("e", _end_block - 1)));
+   bool discover_and_check_last_block_ok(bool is_pruned) {
+      try {
+         //fetch the last block header from the log solely using the log (i.e. not the index: so don't use get_pos()). This is a sanity check.
+         log_header last_header = log.unpack_from<decltype(last_header)>(log.unpack_from<uint64_t>(log.size() - sizeof(uint64_t) - (is_pruned ? sizeof(uint32_t) : 0)));
+         FC_ASSERT(is_ship(last_header.magic) && is_ship_supported_version(last_header.magic), "Unexpected header magic on last block");
+         _end_block    = chain::block_header::num_from_id(last_header.block_id) + 1;
+         last_block_id = last_header.block_id;
+         FC_ASSERT(_begin_block < _end_block, "Block numbers from head and tail of log are not expected");
+      }
+      catch(const std::bad_alloc&) {
+         throw;
+      }
+      catch(const std::exception& e) {
+         ilog("Failure while checking ${name}: ${m}", ("name", log.display_path())("m", e.what()));
+         return false;
+      }
+      return true;
    }
 
-   //only works on non-pruned logs
+   //only works on non-pruned logs since it has to work tail to head
    void recover_blocks() {
-      ilog("recover ${name}.log", ("name", name));
-      uint64_t pos       = 0;
-      uint32_t num_found = 0;
-      log.seek_end(0);
-      const size_t size = log.tellp();
+      const size_t size         = log.size();
+      const size_t header_size  = fc::raw::pack_size(log_header());
+      size_t pos                = 0;
+      uint32_t num_found        = 0;
 
       while (true) {
-         state_history_log_header header;
-         if (pos + state_history_log_header_serial_size > size)
+         if(pos + header_size > size)
             break;
-         log.seek(pos);
-         read_header(header, false);
+         log_header header = log.unpack_from<decltype(header)>(pos);
+
          uint64_t suffix;
-         if (!is_ship(header.magic) || !is_ship_supported_version(header.magic) || header.payload_size > size ||
-             pos + state_history_log_header_serial_size + header.payload_size + sizeof(suffix) > size) {
+         if(!is_ship(header.magic) || !is_ship_supported_version(header.magic) || header.payload_size > size ||
+             pos + header_size + header.payload_size + sizeof(suffix) > size) {
             EOS_ASSERT(!is_ship(header.magic) || is_ship_supported_version(header.magic), chain::plugin_exception,
```

Comment:
The checks are very subtly different,
```c++
        if(!is_ship(header.magic) || !is_ship_supported_version(header.magic) //...
EOS_ASSERT(!is_ship(header.magic) || is_ship_supported_version(header.magic)  //...
```
since this isn't newly added, I'd suggest we noodle on it further before changing it. The `||` in the assert alone is a little perplexing to me. Since `recover_blocks()` is intended to sniff out a corrupted log there might be some nuance to the check here.